Bugfix: Windows compatibility issue with 'cp' command not found when running pnpm start. (#25670)

gin-melodic opened the pull request:
…ing pnpm start. (#25670)

- Replace the existing start script in package.json with a reference to a new copy-and-start.mjs file
- Add new copy-and-start.mjs script to handle copying static files and starting the server
- Improve error handling and logging for the start process
- Make the script more configurable with environment variables

> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 1. Ensure there is an associated issue and you have been assigned to it
> 1. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary

This PR addresses issue #25670 which fixes a Windows compatibility problem where the 'cp' command was not found when running `pnpm start`. The issue occurred because the original start script in package.json used Unix-specific `cp` commands that are not available on Windows systems.

To resolve this cross-platform compatibility issue, I've replaced the simple shell commands with a Node.js script (`copy-and-start.mjs`) that performs the same file copying operations using Node's built-in `fs` module, which works consistently across all platforms.

The new script:
1. Copies static files from `.next/static` to `.next/standalone/.next/static`
2. Copies public resources from `public` to `.next/standalone/public`
3. Starts the server with configurable host and port via environment variables
4. Includes proper error handling and logging

This change ensures that Dify web application can be built and started successfully on both Unix-like systems and Windows systems without requiring additional tools or shell compatibility layers.

Fixes #25670

## Screenshots

| Before | After |
|--------|-------|
| <img width="1088" height="172" alt="image" src="https://github.com/user-attachments/assets/b440f1b5-4b50-4f13-a417-0ea27bff59b2" /> |  <img width="607" height="274" alt="image" src="https://github.com/user-attachments/assets/0149c2aa-92b8-4d4e-8e64-dbab80dc1ab0" />  |

(Note: These are representative images showing the error that was occurring and the successful result after the fix)

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

gin-melodic committed:
bugfix: Windows compatibility issue: 'cp' command not found when running pnpm start. (#25670)

- Replace the existing start script in package.json with a reference to a new copy-and-start.mjs file
- Add new copy-and-start.mjs script to handle copying static files and starting the server
- Improve error handling and logging for the start process
- Make the script more configurable with environment variables

asukaminato0721 commented on web/scripts/copy-and-start.mjs:
```diff
@@ -0,0 +1,74 @@
+if (existsSync('.next/static'))
+  copyDir('.next/static', staticDestDir)
+else
+  console.log('Warning: .next/static directory does not exist')
```
```suggestion
  console.warn('Warning: .next/static directory does not exist')
```

gin-melodic committed:
Address reviewer feedback: improve code maintainability and cross-platform compatibility(#25672)

- Introduce configuration object for directory copy operations
- Add support for copying multiple directories
- Improve logging consistency using console.info and console.warn
- Use path.join for constructing file paths
- Refactor code to be more modular and reusable

gin-melodic commented on web/scripts/copy-and-start.mjs:
```diff
@@ -0,0 +1,74 @@
+if (existsSync('public'))
+  copyDir('public', publicDestDir)
+else
+  console.log('Warning: public directory does not exist')
```
Use path.join: Constructing paths with path.join ensures cross-platform compatibility. 
Good suggestions, I have updated the code accordingly.

Centralize path definitions: Grouping all path-related constants at the top makes the script easier to read and modify.
Good suggestions, I have updated the code accordingly.

Don't Repeat Yourself (DRY): The logic for copying directories is repeated. This can be refactored into a loop over a configuration array, which makes the code cleaner and easier to extend.
Good suggestions, I have updated the code accordingly.

Use braces for control flow: Adding curly braces to if/else statements is a good practice to improve readability and prevent potential bugs.
Because of the `'curly': ['error', 'multi-or-nest', 'consistent'],` rule in `eslint.config.mjs`, I could not add curly braces to the if/else statements.

gin-melodic commented on web/scripts/copy-and-start.mjs:
```diff
@@ -0,0 +1,74 @@
+if (existsSync('public'))
+  copyDir('public', publicDestDir)
+else
+  console.log('Warning: public directory does not exist')
```
Good suggestions, I have updated the code accordingly.

asukaminato0721 commented on web/scripts/copy-and-start.mjs:
```diff
@@ -0,0 +1,79 @@
+ * It is intended to be used as a replacement for `next start`.
+ */
+
+import { cpSync, existsSync, mkdirSync } from 'node:fs'
```
How about fs/promises api ?

gin-melodic replied:
The file operations in the code have been replaced with the `fs/promises api` version

gin-melodic committed:
Rewrite copy-and-start script using async/await

- Replace synchronous FS operations with asynchronous ones
- Improve error handling and logging
- Refactor code into modular functions
- Update server startup process

crazywoola committed:
Merge branch 'main' into main

gin-melodic committed:
Merge branch 'main' into main

gin-melodic committed:
Merge branch 'main' into main

crazywoola committed:
Merge branch 'main' into main

gin-melodic committed:
Merge branch 'main' into main

gin-melodic committed:
Merge branch 'main' into main

gin-melodic committed:
Merge branch 'main' into main

gin-melodic committed:
Merge branch 'main' into main

crazywoola committed:
Merge branch 'main' into main

asukaminato0721 commented:
/gemini review

gin-melodic committed:
Merge remote-tracking branch 'offical/main' and fix review issues.

# Conflicts:
#	web/package.json

gin-melodic committed:
Merge branch 'main' into main

lyzno1 commented on web/package.json:
```diff
@@ -22,7 +22,7 @@
     "build": "next build",
     "build:docker": "next build && node scripts/optimize-standalone.js",
-    "start": "cp -r .next/static .next/standalone/.next/static && cp -r public .next/standalone/public && cross-env PORT=$npm_config_port HOSTNAME=$npm_config_host node .next/standalone/server.js",
+    "start": "node scripts/copy-and-start.mjs",
```
```suggestion
    "start": "node ./scripts/copy-and-start.mjs",
```

gin-melodic committed:
fix(style): style/no-trailing-spaces

lyzno1 commented on web/scripts/copy-and-start.mjs:
```diff
@@ -0,0 +1,112 @@
+        await copyDir(src, dest)
+      else
+        console.warn(`Warning: ${src} directory does not exist`)
+    }
```
[P1] Fail fast when build artifacts missing
    In the previous start script the cp -r .next/static … command aborted the process whenever the build artifacts were missing. In the new script we only log a warning when pathExists(src) is false and
  proceed to start the server anyway. If .next/static (or even public) is absent—e.g. when someone forgets to run pnpm build, or the build step fails—the server now boots successfully but serves none of
  the static assets, resulting in a blank/broken UI instead of a hard failure. Could we keep the old fail-fast behavior by exiting with a non-zero status when these directories are missing?

gin-melodic committed:
Merge branch 'langgenius:main' into main

gin-melodic committed:
fix: improve error handling in copy-and-start script

Updated package.json start script to use explicit relative path for clarity. Enhanced copy-and-start.mjs to treat missing directories as critical errors, logging detailed messages and exiting the process to ensure build reliability and prevent silent failures.
